File: src/api/aiInvocation.ts
/**
 * AI 调用审计 API 模块
 *
 * 对接后端 /api/v1/ai-invocation 端点，提供成本聚合与调用记录查询。
 * 成本单位统一为人民币元（与 ApiCostLog 一致），后端返回 cost_usd 字段
 * 在前端展示时按汇率转换为人民币元。
 */
import request from '@/utils/request'

// ============== 类型定义 ==============

/** 批次成本汇总信息 */
export interface BatchCostInfo {
  /** 总调用次数 */
  totalCalls: number
  /** 总输入 Token 数 */
  totalPromptTokens: number
  /** 总输出 Token 数 */
  totalCompletionTokens: number
  /** 总 Token 数（输入 + 输出） */
  totalTokens: number
  /** 估算成本（美元） */
  totalCostUsd: number
}

/** AI 调用成本聚合查询参数 */
export interface AIInvocationStatsParams {
  project_id: number
  start_date?: string
  end_date?: string
  group_by?: 'model' | 'strategy' | 'date'
}

/** AI 调用成本聚合结果项 */
export interface AIInvocationStatsItem {
  group_key: string
  total_calls: number
  total_prompt_tokens: number
  total_completion_tokens: number
  total_cost_usd: number
}

/** AI 调用成本聚合响应 */
export interface AIInvocationStatsResponse {
  items: AIInvocationStatsItem[]
}

/** AI 调用记录查询参数 */
export interface AIInvocationListParams {
  batch_id?: number
  project_id?: number
  page?: number
  page_size?: number
}

/** AI 调用记录列表项 */
export interface AIInvocationListItem {
  id: number
  run_id: number | null
  step_name: string | null
  model: string
  prompt_tokens: number
  completion_tokens: number
  cost_usd: number
  latency_ms: number
  status: string
  error_message: string | null
  created_at: string
  generation_batch_id: number | null
  scenario_type: string | null
  generation_strategy: string | null
  prompt_key: string | null
  prompt_version: number | null
  prompt_hash: string | null
  error_code: string | null
}

/** AI 调用记录分页响应 */
export interface AIInvocationListResponse {
  items: AIInvocationListItem[]
  total: number
  page: number
  page_size: number
}

// ============== 工具函数 ==============

/** USD 转 CNY 汇率常量（与 ApiCostLog 人民币元对齐） */
const USD_TO_CNY = 7.25

/** 将 USD 金额转换为人民币元 */
export function usdToCny(usd: number): number {
  return Number((usd * USD_TO_CNY).toFixed(4))
}

/** 从响应中提取 data 字段 */
function extractData<T>(response: unknown): T {
  const resp = response as { code?: number; data?: T }
  if (resp && resp.data !== undefined) return resp.data
  return response as T
}

// ============== API 封装 ==============

export const aiInvocationApi = {
  /**
   * 按批次 ID 查询成本汇总
   *
   * 通过 /list 接口获取该批次所有调用记录，
   * 在前端汇总 Token 数量和成本。
   *
   * @param batchId - 生成批次 ID
   * @returns 成本汇总信息
   */
  getBatchCost: async (batchId: number): Promise<BatchCostInfo> => {
    const response = await request.get('/api/v1/ai-invocation/list', {
      params: { batch_id: batchId, page: 1, page_size: 100 },
    })
    const data = extractData<AIInvocationListResponse>(response)
    const items = data.items || []

    let totalPromptTokens = 0
    let totalCompletionTokens = 0
    let totalCostUsd = 0

    for (const item of items) {
      totalPromptTokens += item.prompt_tokens || 0
      totalCompletionTokens += item.completion_tokens || 0
      totalCostUsd += item.cost_usd || 0
    }

    return {
      totalCalls: items.length,
      totalPromptTokens,
      totalCompletionTokens,
      totalTokens: totalPromptTokens + totalCompletionTokens,
      totalCostUsd,
    }
  },

  /** 获取 AI 调用成本聚合统计 */
  getStats: async (
    params: AIInvocationStatsParams
  ): Promise<{ code: number; message: string; data: AIInvocationStatsResponse }> => {
    return request.get('/api/v1/ai-invocation/stats', { params })
  },

  /** 获取 AI 调用记录分页列表 */
  getList: async (
    params: AIInvocationListParams
  ): Promise<{ code: number; message: string; data: AIInvocationListResponse }> => {
    return request.get('/api/v1/ai-invocation/list', { params })
  },
}

export default aiInvocationApi

File: app/schemas/ai_invocation.py
"""
AI 调用审计查询 Schema 模块

本模块定义 AIInvocationGateway 审计增强相关的请求/响应 Schema：
    - AIInvocationStatsQuery: 成本聚合查询参数
    - AIInvocationStatsResponse: 成本聚合结果项
    - AIInvocationListQuery: 调用记录查询参数
    - AIInvocationListItem: 调用记录列表项
"""
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


class AIInvocationStatsQuery(BaseModel):
    """AI 调用成本聚合查询参数

    Attributes:
        project_id: 项目ID（必填）
        start_date: 起始日期（含），默认近30天
        end_date: 截止日期（含），默认今天
        group_by: 聚合维度：model/strategy/date
    """
    project_id: int = Field(..., gt=0, description="项目ID")
    start_date: Optional[date] = Field(None, description="起始日期（含）")
    end_date: Optional[date] = Field(None, description="截止日期（含）")
    group_by: str = Field("model", pattern="^(model|strategy|date)$", description="聚合维度")


class AIInvocationStatsResponse(BaseModel):
    """AI 调用成本聚合结果项

    Attributes:
        group_key: 聚合键值（模型名/策略名/日期字符串）
        total_calls: 总调用次数
        total_prompt_tokens: 总输入 Token 数
        total_completion_tokens: 总输出 Token 数
        total_cost_usd: 总成本（美元）
    """
    group_key: str
    total_calls: int
    total_prompt_tokens: int
    total_completion_tokens: int
    total_cost_usd: float


class AIInvocationListQuery(BaseModel):
    """AI 调用记录查询参数

    Attributes:
        batch_id: 生成批次ID（可选）
        project_id: 项目ID（可选）
        page: 页码，从1开始
        page_size: 每页条数
    """
    batch_id: Optional[int] = Field(None, description="生成批次ID")
    project_id: Optional[int] = Field(None, gt=0, description="项目ID")
    page: int = Field(1, ge=1, description="页码")
    page_size: int = Field(20, ge=1, le=100, description="每页条数")


class AIInvocationListItem(BaseModel):
    """AI 调用记录列表项

    包含 AICallLog 全部字段（含审计增强字段），
    通过 generation_batch → project 间接关联项目。

    Attributes:
        id: 记录ID
        run_id: 关联运行ID
        step_name: Step 名称
        model: 使用的AI模型
        prompt_tokens: 输入Token数
        completion_tokens: 输出Token数
        cost_usd: 调用成本（美元）
        latency_ms: 调用耗时（毫秒）
        status: 状态
        error_message: 错误信息
        created_at: 创建时间
        generation_batch_id: 关联生成批次ID
        scenario_type: 资料组合场景
        generation_strategy: 生成策略
        prompt_key: Prompt模板键
        prompt_version: Prompt版本号
        prompt_hash: Prompt内容哈希
        error_code: 错误码枚举值
    """
    id: int
    run_id: Optional[int] = None
    step_name: Optional[str] = None
    model: str
    prompt_tokens: int
    completion_tokens: int
    cost_usd: float
    latency_ms: int
    status: str
    error_message: Optional[str] = None
    created_at: datetime
    generation_batch_id: Optional[int] = None
    scenario_type: Optional[str] = None
    generation_strategy: Optional[str] = None
    prompt_key: Optional[str] = None
    prompt_version: Optional[int] = None
    prompt_hash: Optional[str] = None
    error_code: Optional[str] = None

    model_config = {"from_attributes": True}
